File: scraper/src/common/remove-tag-duplicates.ts
import SQL from '@nearform/sql';
import _ from 'lodash';

import database from './database/index.js';
import logger from './logger.js';
import sequentialAsyncForEach from './sequential-async-for-each.js';

const { map, replace } = _;

export default async function removeTagDuplicates(
  tagName: string,
): Promise<string[]> {
  logger.debug(`removeTagDuplicates(${tagName})`);
  const query = SQL`
    SELECT "weighted_tags"."name" AS "name"
    FROM (
      SELECT
        "Tag"."albumsScrapedAt",
        "Tag"."listCheckedAt",
        "Tag"."listUpdatedAt",
        "Tag"."name",
        "Tag"."registeredAt",
        SUM(
          "AlbumTag"."count" :: FLOAT * COALESCE("Album"."playcount", 0) / 1000000 * COALESCE("Album"."listeners", 0) / 1000
        ) AS "weight"
      FROM "Tag"
      JOIN "AlbumTag" ON "AlbumTag"."tagName" = "Tag"."name"
      JOIN "Album" ON "Album"."artist" = "AlbumTag"."albumArtist"
      AND "Album"."name" = "AlbumTag"."albumName"
      WHERE
        REGEXP_REPLACE("Tag"."name", '\\W+', '', 'g') = ${replace(
          tagName,
          /[^\da-z]/gi,
          '',
        )}
      GROUP BY "Tag"."name"
      ORDER BY "weight" DESC
    ) AS "weighted_tags" OFFSET 1
  `;

  // logger.debug(query);
  const result = await database.query<{ name: string }>(query);
  await sequentialAsyncForEach(result.rows, async ({ name }) => {
    await database.query(SQL`
      DELETE FROM "Tag"
      WHERE "name" = ${name}
    `);
  });
  return map(result.rows, 'name');
}

File: scraper5/utils/data/removeTagDuplicates.go
package data

import (
	"lac/scraper5/database"
	"lac/scraper5/utils"
	"log"
)

func RemoveTagDuplicates(tagName string) {
	rows, err := database.Database.Query(`SELECT "weighted_tags"."name" AS "name"
  FROM (
    SELECT
      "Tag"."albumsScrapedAt",
      "Tag"."listCheckedAt",
      "Tag"."listUpdatedAt",
      "Tag"."name",
      "Tag"."registeredAt",
      SUM(
        "AlbumTag"."count" :: FLOAT * COALESCE("Album"."playcount", 0) / 1000000 * COALESCE("Album"."listeners", 0) / 1000
      ) AS "weight"
    FROM "Tag"
    JOIN "AlbumTag" ON "AlbumTag"."tagName" = "Tag"."name"
    JOIN "Album" ON "Album"."artist" = "AlbumTag"."albumArtist"
    AND "Album"."name" = "AlbumTag"."albumName"
    WHERE
      REGEXP_REPLACE("Tag"."name", '\\W+', '', 'g') = $1
    GROUP BY "Tag"."name"
    ORDER BY "weight" DESC
  ) AS "weighted_tags" OFFSET 1`, utils.RejectNonAlphanumeric(tagName))
	utils.HandleError(err, "utils/data/removeTagDuplicates.go:RemoveTagDuplicates:query")
	defer rows.Close()
	tagNamesToRemove := []string{}
	for rows.Next() {
		var tagNameToRemove string
		err = rows.Scan(&tagNameToRemove)
		utils.HandleError(err, "utils/data/removeTagDuplicates.go:RemoveTagDuplicates:scan")
		tagNamesToRemove = append(tagNamesToRemove, tagNameToRemove)
	}
	for _, tagNameToRemove := range tagNamesToRemove {
		log.Printf("Removing duplicate tag %s; primary tag %s", tagNameToRemove, tagName)
		DeleteTag(tagNameToRemove)
	}
}
